bin/build-dist

```diff
@@ -2,32 +2,86 @@
 
 set -eux
 
+PREFER_AUTHORING=${PREFER_AUTHORING:='1'}
 BUILD_DIR=${BUILD_DIR:='build_dir'}
+INSTALL_DIR=
+
+case "$1" in
+	--favour-authoring | --prefer-authoring | --authoring)
+		PREFER_AUTHORING="1"
+		;;
+	--disfavour-authoring | --prefer-builder | --builder )
+		PREFER_AUTHORING="0"
+		;;
+esac
+
+function error () {
+	echo >&2 "$@"
+	return 1
+}
+
+function last-modified-dir () {
+    ls -td -- */ | head -n 1 | cut -d'/' -f1
+}
+
+function module-build-install-dir () {
+	perl -MModule::Build -e'print Module::Build->resume->dist_dir;'
+}
+
+function authoring-tool-favoured () {
+	case "$FAVOUR_AUTHORING" in
+		0 | false) return 0 ;;
+		1 | true ) return 1 ;;
+		*)
+			error "Unknown value of FAVOUR_AUTHORING - '$FAVOUR_AUTHORING'"
+			;;
+	esac
+}
+
+function build-with-dist-zilla () {
+	[ -e dist.ini ] || return 1
+	[ dzil -v >/dev/null 2>&1 ] || return 1
+
+	dzil authordeps --missing | cpm install -
+	dzil build --no-tgz --in "$BUILD_DIR"
+}
+
+function build-with-minilla () {
+	[ -e minil.toml ] || return 1
+	[ minil -v >/dev/null 2>&1 ] || return 1
 
-if [[ -e dist.ini ]]; then
-    dzil build --no-tgz --in "$BUILD_DIR"
-elif [[ -e minil.toml ]]; then
-    rm -rf "$BUILD_DIR"
     minil build
 
     # TODO: It looks like minil doesn't accept a build directory name, so we
     # just find the last created dir and use that.
     # shellcheck disable=SC2012
-    DIR=$(ls -td -- */ | head -n 1 | cut -d'/' -f1)
-    mv "$DIR" "$BUILD_DIR"
-elif [[ -e Build.PL ]]; then
-    rm -rf "$BUILD_DIR"
+    INSTALL_DIR=$(last-modified-dir)
+}
+
+function build-with-authoring-tool () {
+	# false used as diff-friendly syntax sugar
+	false                           \
+		|| build-with-dist-zilla    \
+		|| build-with-minilla       \
+		|| false
+}
+
+function build-with-authoring-tool-favoured () {
+	authoring-tool-favoured && build-with-authoring-tool
+}
+
+function build-with-build-pl () {
     perl Build.PL
     ./Build clean
     ./Build manifest
     ./Build distdir
 
-    # Module::Build does not support a build directory, so we have to find
-    # the last created dir and use that.
-    DIR="$(perl -MModule::Build -e'print Module::Build->resume->dist_dir;')"
-    mv "$DIR" "$BUILD_DIR"
-elif [[ -e Makefile.PL ]]; then
-    rm -rf "$BUILD_DIR"
+    # Module::Build does not support a build directory, so we have to find it
+	# Module::Build::Tiny doesn't support `resume` so fallback to last modified
+    INSTALL_DIR="$( module-build-install-dir || last-modified-dir )"
```

Comment:
You are using both spaces and tabs for indent here. Consider 2 space indent :)

Reply:
talking style before functionality, contributor unfriendly environment - these days anyone can set tab to its preferred value, so everyone can use what they are used to, that's why I rarely contribute anymore

Reply:
@happy-barney he's just trying to be helpful. We can set up a formatter for this. It's not a blocker on the PR.

Reply:
I'm just upset and demotivated. Plus original script has indentation 4 spaces ... I will not mind such comments once we agree on functionality (and I will ignore last one) but now it triggered my fight-or-flight in favour of later.

Reply:
For context, @ingydotnet is really into `bash` and was just excited to share some tips. :) I asked him to check on this since he really likes this stuff. So, I think it would be helpful to look at it as a well-meaning peer review and not a criticism of your work.

There have been very few people who have taken an interest in touching this code, so I really appreciate that you've taken the time to do so.

One concern that I had was if we're going to be make it much more flexible, it would be helpful to try to hash out an API that won't leave us boxed in at some future point. I don't have a deep enough understanding of `bash` to be really helpful in that way.

I think he's going to come up with a slightly different approach for this and we'll see where we land at the end of the day, but the last thing I want to see is anyone being demotivated or upset because of code review comments here. It's open source and we should all be having fun. 😄

Reply:
anyway, what I really need is to be able to prefer builders over authoring tools. That will be enough (out of everything else in this PR)

What I'm more concerned reusable workflows (would be nice to have them under `perl-actions`) and related actions.

Reply:
@happy-barney Sorry I got mixed up on what was going on. I shouldn't have done the review that way.
The functionality and the bash cleanup should be done separately.
Is there a way for me to remove my review?

Reply:
@ingydotnet no problem, let's keep it, it will come anyway :-) I should also change this PR to draft.